Ensure tilde$1 onExit is run in correct order

masonedmison opened the pull request:
Resolves #13049

masonedmison commented on instrumentation/pekko/pekko-http-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/pekkohttp/v1_0/server/route/RouteConcatenationInstrumentation.java:
```diff
@@ -39,8 +58,10 @@ public static void onEnter() {
+    public static void onExit(
+        @Advice.Argument(value = 2) RequestContext ctx,
+        @Advice.Return(readOnly = false) Future<RouteResult> fut) {
+      fut = fut.andThen(new OnExitFinalizer(), ctx.executionContext());
```
I need to have a deeper think about the choice of `andThen` here--`andThen` will swallow any exceptions that occur within the PartialFunction which I don't think is what we want.

masonedmison replied:
Actually, exceptions raised within the future will still be logged by the Exception Logger. Resolving this.

trask commented:
how hard would it be to add a test for this? thanks

trask commented:
(if you rebase, we've fixed the CI failure that you're seeing)

masonedmison commented:
> how hard would it be to add a test for this? thanks

I'll give it a try. Probably a stupid question, but I'm not very familiar with gradle. How does one run tests for a particular module? 

The docs say this
```
### Executing single test

Executing `./gradlew :instrumentation:<INSTRUMENTATION_NAME>:test --tests <TEST FILE NAME>` will run only the selected test.
```

but that doesn't seem to work.

masonedmison commented:
> how hard would it be to add a test for this? thanks

@trask I just played around with this for a bit and I'm having a difficult time reproducing this using straight pekko code; it might be difficult to write a failing test without introducing the tapir-pekko module.

laurit commented on instrumentation/pekko/pekko-http-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/pekkohttp/v1_0/server/route/RouteConcatenationInstrumentation.java:
```diff
@@ -27,6 +33,19 @@ public void transform(TypeTransformer transformer) {
         named("$anonfun$$tilde$2"), this.getClass().getName() + "$Apply2Advice");
   }
 
+  public static class OnExitFinalizer implements PartialFunction<Try<RouteResult>, Unit> {
+    @Override
+    public boolean isDefinedAt(Try<RouteResult> x) {
+      return true;
+    }
+
+    @Override
+    public Unit apply(Try<RouteResult> v1) {
+      PekkoRouteHolder.restore();
+      return null;
+    }
+  }
```
Could you move this class to a top level class or to some other util class. We usually don't add anything besides the advice classes here.

laurit commented on instrumentation/pekko/pekko-http-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/pekkohttp/v1_0/server/route/RouteConcatenationInstrumentation.java:
```diff
@@ -39,8 +58,10 @@ public static void onEnter() {
+    public static void onExit(
+        @Advice.Argument(value = 2) RequestContext ctx,
+        @Advice.Return(readOnly = false) Future<RouteResult> fut) {
+      fut = fut.andThen(new OnExitFinalizer(), ctx.executionContext());
```
This advice will apply also when the original method threw an exception because of the `onThrowable = Throwable.class`. Perhaps it would be better to write this as
```
    public static void onExit(
        @Advice.Argument(value = 2) RequestContext ctx,
        @Advice.Return(readOnly = false) Future<RouteResult> future,
        @Advice.Thrown Throwable throwable) {
  if (throwable != null) {
    PekkoRouteHolder.restore();
  } else {
    future = future.andThen(new OnExitFinalizer(), ctx.executionContext());
  }
```
I don't know how scala futures work but in java `CompletableFuture` the methods starting with `then` only apply when the future completes successfully. With `CompletableFuture` `whenComplete` would be a better choice. Please check whether `andThen` also runs when the future does not complete normally.

masonedmison replied:
`andThen` will be evaluated regardless of how the `Future` completes: the input of this supplied function to `andThen` takes a `Try` which models both success and failures, and the function we supply executes `PekkoRouteHolder.restore()` regardless of the input; e.g., we unconditionally restore regardless of how the `Future` completed. If I understand the above correctly, I believes this does what we want.

laurit replied:
Thanks for confirming this.

laurit commented:
> > how hard would it be to add a test for this? thanks
> 
> @trask I just played around with this for a bit and I'm having a difficult time reproducing this using straight pekko code; it might be difficult to write a failing test without introducing the tapir-pekko module as it's only with [this](https://github.com/softwaremill/tapir/blob/master/server/pekko-http-server/src/main/scala/sttp/tapir/server/pekkohttp/PekkoHttpServerInterpreter.scala#L42) construction of the route that the issue is triggered (at least as far I've experienced).

If you are reluctant to add extra dependencies you could add a separate test suite (search for `registering(JvmTestSuite::class)`) Suites have their own set of dependencies and are isolated from the other tests.

masonedmison commented:
> > > how hard would it be to add a test for this? thanks
> > 
> > 
> > @trask I just played around with this for a bit and I'm having a difficult time reproducing this using straight pekko code; it might be difficult to write a failing test without introducing the tapir-pekko module as it's only with [this](https://github.com/softwaremill/tapir/blob/master/server/pekko-http-server/src/main/scala/sttp/tapir/server/pekkohttp/PekkoHttpServerInterpreter.scala#L42) construction of the route that the issue is triggered (at least as far I've experienced).
> 
> If you are reluctant to add extra dependencies you could add a separate test suite (search for `registering(JvmTestSuite::class)`) Suites have their own set of dependencies and are isolated from the other tests.

@laurit I ended up adding the necessary dependency as a `testImplementation`. With this, I was able to write a failing test (e.g., I added the test and dropped the commit that modifies the `onExit` logic, and the test failed; with the commit back in place, the test passes).

masonedmison commented:
@trask Do you have a sense of when this will be released?
